src/main/java/org/jooq/lambda/concurrent/SameExecutorCompletionStage.java

```diff
@@ -0,0 +1,222 @@
+package org.jooq.lambda.concurrent;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+final class SameExecutorCompletionStage<T> implements CompletionStage<T> {
+    private final CompletionStage<T> delegate;
+    private final Executor defaultExecutor;
```

Comment:
Hmm, I hadn't thought of allowing `null` in the public API... But there could be `Async.supplyAsync(() -> "test"))` which would internally set the `Executor` to `null` and then not specify any executor when delegating to the `delegate`...

I'm not sure if we should allow `null` in the public API, so I'd also prefer 3)